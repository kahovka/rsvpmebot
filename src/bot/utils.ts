import TelegramBot from 'npm:node-telegram-bot-api';
import { match } from 'npm:ts-pattern@^5.6.2';
import { eventCollection } from '../db/mongo.ts';
import { RSVPEvent, RSVPEventState } from '../db/types.ts';
import { logger } from '../logger.ts';
import {
	settingDescriptionState,
	settingNameState,
	settingParticipantLimitState
} from './botStates.ts';

export const getEventDescriptionHtml = (event: RSVPEvent) => {
	return event.name + '\n' + event.description + '\n' + `Participants: ${event.participantLimit}`;
};

export const getEventNextState = (event: RSVPEvent): RSVPEventState => {
	return match(event.state)
		.with(RSVPEventState.NewEvent, () => settingNameState.nextState)
		.with(RSVPEventState.NameSet, () => settingDescriptionState.nextState)
		.with(RSVPEventState.DescriptionSet, () => settingParticipantLimitState.nextState)
		.otherwise(() => RSVPEventState.NewEvent);
};

export const botActionErrorCallback = (
	error: unknown,
	bot: TelegramBot,
	message: TelegramBot.Message
) => {
	logger.error('Failed ({error}) processing message: {message}', { message, error });
	bot.sendMessage(
		message.chat.id,
		'Something went wrong, please try to response to the previous bot message again or start anew'
	);
};

export const deleteExistingMessagesAndReply = async (
	bot: TelegramBot,
	message: TelegramBot.Message,
	event: RSVPEvent,
	messageToSend: string,
	replyMarkup: string
) => {
	await bot.deleteMessage(message.chat.id, event.lastMessageId);
	await bot.deleteMessage(message.chat.id, message.message_id);

	await bot
		.sendMessage(message.chat.id, messageToSend, {
			...(replyMarkup && { reply_markup: replyMarkup }),
			parse_mode: 'HTML'
		})
		.then((replyMessage: TelegramBot.Message) => {
			eventCollection().updateOne(
				{ _id: event._id },
				{
					$set: {
						state: getEventNextState(event),
						lastMessageId: replyMessage.message_id
					}
				},
				{ upsert: true }
			);
		});
};
