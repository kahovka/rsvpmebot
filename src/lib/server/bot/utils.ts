import TelegramBot from 'node-telegram-bot-api';
import { match } from 'ts-pattern';
import { eventCollection } from '../db/mongo.ts';
import { RSVPEventParticipant, RSVPEvent, RSVPEventState } from '../db/types.ts';
import { logger } from '../../../logger.ts';
import {
	settingDescriptionState,
	settingNameState,
	settingParticipantLimitState
} from './botStates.ts';

export const getParticipantDisplayName = (participant: RSVPEventParticipant) =>
	`${participant.firstName} (${participant.username})`;

export const getEventDescriptionHtml = (event: RSVPEvent) => {
	const allParticipants = event.participantsList
		?.map((participant) => getParticipantDisplayName(participant))
		?.join('<br>');
	const allWaiting = event.waitlingList
		?.map((participant) => getParticipantDisplayName(participant))
		?.join('<br>');
	return `
		<p>${event.name ?? 'Your event'}</p>
		<p><b>Description:</b> ${event.description ?? 'Your event description'}</p>
		<p><b>Number of participants:</b> ${event.participantLimit}</p>
		${event.participantsList && event.participantsList.length > 0 ? '<p><b>Participants:</b></p>' + allParticipants : ''}
		${event.waitlingList && event.waitlingList.length > 0 ? '<p><b>Waiting:</b></p>' + allWaiting : ''}
		`;
};

export const getEventNextState = (event: RSVPEvent): RSVPEventState => {
	return match(event.state)
		.with(RSVPEventState.NewEvent, () => settingNameState.nextState)
		.with(RSVPEventState.NameSet, () => settingDescriptionState.nextState)
		.with(RSVPEventState.DescriptionSet, () => settingParticipantLimitState.nextState)
		.otherwise(() => RSVPEventState.NewEvent);
};

export const botActionErrorCallback = (
	error: unknown,
	bot: TelegramBot,
	message: TelegramBot.Message
) => {
	logger.error('Failed ({error}) processing message: {message}', { message, error });
	bot.sendMessage(
		message.chat.id,
		'Something went wrong, please try to response to the previous bot message again or start anew'
	);
};

export const deleteExistingMessagesAndReply = async (
	bot: TelegramBot,
	message: TelegramBot.Message,
	event: RSVPEvent,
	messageToSend: string,
	replyMarkup: string
) => {
	await bot.deleteMessage(message.chat.id, event.lastMessageId);
	await bot.deleteMessage(message.chat.id, message.message_id);

	await sendNewEventMessage(bot, message.chat.id, event, messageToSend, replyMarkup);
};

export const sendNewEventMessage = async (
	bot: TelegramBot,
	chatId: number,
	event: RSVPEvent,
	messageToSend: string,
	replyMarkup: string
) => {
	await bot
		.sendMessage(chatId, messageToSend, {
			...(replyMarkup && { reply_markup: replyMarkup }),
			parse_mode: 'HTML'
		})
		.then((replyMessage: TelegramBot.Message) => {
			eventCollection().updateOne(
				{ _id: event._id },
				{
					$set: {
						state: getEventNextState(event),
						lastMessageId: replyMessage.message_id
					}
				},
				{ upsert: true }
			);
		});
};
