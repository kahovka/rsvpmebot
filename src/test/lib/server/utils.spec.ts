import { describe, expect, test } from 'vitest';
import { getEventDescriptionHtml } from '../../../lib/server/bot/utils.ts';
import { fakeStaticCompletedEvent } from '../../fixtures/eventsFixtures.ts';

describe('utils tests', () => {
	test('simple test', () => {
		expect(getEventDescriptionHtml(fakeStaticCompletedEvent)).toMatchSnapshot();
	});
});
