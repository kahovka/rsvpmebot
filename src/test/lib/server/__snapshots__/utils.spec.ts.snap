// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils tests > simple test 1`] = `
"
		<p>indeed garrote</p>
		<p><b>Description:</b> joyfully hence negative augment republican early sentimental generously from underachieve usually fashion if around famously</p>
		<p><b>Number of participants:</b> 3</p>
		<p><b>Participants:</b></p>Isom (@Breitenberg)<br>Willy (@Hyatt)<br>Alphonso (@Schulist)
		<p><b>Waiting:</b></p>Winifred (@Wyman)<br>Lucious (@Labadie)
		"
`;
